doc/error-handling/error_handling_design_spec.md

```diff
@@ -0,0 +1,429 @@
+
+## 3.2 Error codes
+As per SONIC architecture, applications do not invoke SAI APIs directly and they are not aware of underlying SAI error codes. A new set of generic error codes are defined for applications, which are internally mapped to underlying SAI error codes.
+
+As SWSS submodule already provides a common library accessible to all SONIC applications, a new set of common error codes are defined as part of this library (applications need not include any new header files). The following table shows the mapping of SWSS error codes to SAI error codes.
+
+| SWSS error code       | SAI error code                 |
+|-----------------------|--------------------------------|
+| SWSS_RC_SUCCESS       | SAI_STATUS_SUCCESS             |
+| SWSS_RC_INVALID_PARAM | SAI_STATUS_INVALID_PARAMETER   |
+| SWSS_RC_UNAVAIL       | SAI_STATUS_NOT_SUPPORTED       |
+| SWSS_RC_NOT_FOUND     | SAI_STATUS_ITEM_NOT_FOUND      |
+| SWSS_RC_NO_MEMORY     | SAI_STATUS_NO_MEMORY           |
+| SWSS_RC_EXISTS        | SAI_STATUS_ITEM_ALREADY_EXISTS |
+| SWSS_RC_FULL          | SAI_STATUS_TABLE_FULL          |
+| SWSS_RC_IN_USE        | SAI_STATUS_OBJECT_IN_USE       |
+
+## 3.3 OrchAgent changes
+
+OrchAgent is the only component responsible for adding/modifying/deleting objects in ERROR_DB database. Since the error handling framework is part of the orchestration agent, the error notifications are reported only when orchestration agent is running.
+
+OrchAgent defines the following new classes to handle error reporting/listening:
+
+1. **Error reporter class** - defines functions to write to ERROR_DB and publish the notifications.
+2. **Error listener class** - defines functions to register/de-register for notifications on a specific table. Applications need to specify the table name, operation such CREATE/DELETE/UPDATE, notification type (success/failure/both)  and callback function when instantiating the error listener class. Applications can instantiate multiple listeners to receive notifications from different tables.
+
+### 3.3.1 Event processing
+
+Below diagram shows how framework reports notifications to the registered listeners:
+
+![Framework high level design](images/error_handling_overview.png)
+
+The following is the sequence of events involved in reporting a failure notification from Syncd process to application:
+
+1. Syncd reports errors using a single notification channel to OrchAgent. Using a single notification channel ensures that order of the notifications is retained.
+2. After receiving the notification from Syncd, OrchAgent:
+   - Translates it from SAI data types to ERROR_DB data types
+   - Adds an entry in to error database. If the entry already exists, the corresponding failure code is updated.
+   - Publishes the notifications to respective error listeners.
+3. Error listener waits for the incoming notifications, filters them and invokes the application callback.

```

Comment:
Can you describe how does Error listener filters the notifications ? what is the criteria supported? please add the use case?

Reply:
Sure, we will add more details on this.